Dockerfile

```diff
@@ -1,9 +1,13 @@
 FROM alpine:latest
 
+# first add the file because we have to run setcap on the binary and we would not like to create a new image layer so the binary must exist on RUN
+ADD coredns /coredns
+
 # Only need ca-certificates & openssl if want to use DNS over TLS (RFC 7858).
-RUN apk --no-cache add bind-tools ca-certificates openssl && update-ca-certificates
+RUN apk --no-cache add bind-tools ca-certificates openssl libcap && update-ca-certificates && chown nobody:nobody /coredns && setcap CAP_NET_BIND_SERVICE=+eip /coredns
 
-ADD coredns /coredns
+USER nobody
```

Comment:
this works, but now we depend on setcap, and this further complicates building for other architectures.
IMO the fix should be run on a high port as user non-root.

But then we need to port forward which you can't specify in the Dockerfile :-(

Reply:
:smile: that's the problem. Security is always a trade off and most of the time between easy and complicated.

Reply:
What about the idea to create a pipeline which builds the binary and test it in the common way and then have the Dockerfiles for the different images (and different possibilities) in a separate directory in the repository with multiple pipelines for the image builds. Then it would be possible to create secure(er) images at least for some platforms. That won't be a 100 percent solution but anyway better than nothing.

Reply:
Yes, this needs more then just changing the Dockerfile.

Right now we also don't install certs in non-amd64 images which could be fixed in the same way.

I can build those images on my laptop, but that code is non-portable. We looked at docker automatic builds a while back, but nothing came out of it and I'm not sure it can actually do multi arch builds

Reply:
Yes, that's true. I haven't looked at multi arch build too. And I have to say that we use Gitlab (on-prem) and the pipelines there and therefore I have no experience with Travis and the possibilities there.

We sometimes have to build build images. What I mean is, maybe it is possible to put your local build code into a prepared build image and to use this image then to build the coredns release  images. But this needs stacked pipelines. The first one build the build system and the second one build the release images with the result from the first one.

Reply:
for Linux you'll need qemu and binfmt to be setup in your kernel, see https://github.com/miekg/dxbuild

This won't even run in Travis...

Reply:
:smile: That's why we use Gitlab on-prem because than you are able to use Windows with Gitlab runners to run UI tests from the pipeline or to prepare special build hosts to run pipeline steps on them and so on for example. 80% of the builds are perfectly happy with the default solution but the other 20% are special, every single one of them. :rescue_worker_helmet: